src/OpenTelemetry.Api/Trace/TracerProvider.cs

```diff
@@ -43,6 +48,83 @@ protected TracerProvider()
     /// <param name="name">Name identifying the instrumentation library.</param>
     /// <param name="version">Version of the instrumentation library.</param>
     /// <returns>Tracer instance.</returns>
-    public Tracer GetTracer(string name, string? version = null)
-        => new(new ActivitySource(name ?? string.Empty, version));
+    public Tracer GetTracer(
+#if NET6_0_OR_GREATER
+        [AllowNull]
+#endif
+        string name,
+        string? version = null)
+    {
+        var tracers = this.tracers;
+        if (tracers == null)
+        {
+            // Note: Returns a no-op Tracer once dispose has been called.
+            return new(activitySource: null);
+        }
+
+        var key = new TracerKey(name, version);
+
+Retry:
+        if (!tracers.TryGetValue(key, out var tracer))
+        {
+            lock (tracers)
+            {
+                if (this.tracers == null)
+                {
+                    // Note: We check here for a race with Dispose and return a
+                    // no-op Tracer in that case.
+                    return new(activitySource: null);
+                }
+
+                tracer = new(new(key.Name, key.Version));
+                if (!tracers.TryAdd(key, tracer))
+                {
+                    // Note: This should really never happen due to the
```

Comment:
Let's utilize this knowledge and keep this simple. No need to check the return value of `TryAdd`. We do the same in AggregatorStore: https://github.com/open-telemetry/opentelemetry-dotnet/blob/dc1f09d4b968f96955ccfeaa5b0f6f7e9989656a/src/OpenTelemetry/Metrics/AggregatorStore.cs#L504

Reply:
I switched it to a Debug.Assert verification.